File: ports/python/src/unicode_python/security/display/renderer_divergence.py
"""RendererDivergence — detection of codepoint/sequence shapes known to
render differently across font + terminal + browser stacks (the
display-layer detector).

Byte-faithful transliteration of the verified Rust reference
implementation (itself a transliteration of
``Unicode.Security.Display.RendererDivergence``).

Threat model. An adversary crafts content that renders one way in the
auditor's renderer (a benign glyph or an empty span) and a different way
in the consumer's renderer (a misleading glyph, a wider glyph, or a
different sequence). This is the "fingerprint stability" family — clear
inputs render the same across the renderer cohort the Standard documents
as stable.

What the detector draws. A heuristic three-value split, surfaced through
the universal clear/hazard carrier: an input is clear when none of the
documented variance triggers fire, and otherwise is classified by the
first trigger in priority order — combining-mark stack overflow,
variation-selector presence, an unregistered ZWJ shape,
fullwidth/halfwidth display, or mixed direction. It reuses the port's own
tables (the variation-selector set, the grapheme Extend class, the RGI
ZWJ registry, and the strong-bidi classes), never a host rendering or
shaping library.

Sub-threats (priority order):
    1. ``CombiningStackOverflow``    Zalgo-like combining-mark stack >= 4 on a base.
    2. ``VariationSelectorVariance`` any variation selector present.
    3. ``UnregisteredZwjVariance``   ZWJ-containing input not in the RGI ZWJ set.
    4. ``FullwidthVariance``         a fullwidth/halfwidth codepoint present.
    5. ``MixedDirectionVariance``    both strong-LTR and strong-RTL codepoints.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Union

from ..calculus import ClassificationKind
from ..covert import variation_selector_payload
from ..identity import emoji_zwj_integrity, ucd
from ...segmentation import grapheme

# ─────────────────────────────────────────────────────────────────────
# §1 Constants
# ─────────────────────────────────────────────────────────────────────

# The combining-mark stack depth (on a single base) at or beyond which
# the input is treated as a Zalgo-style rendering-variance hazard.
MIN_COMBINING_STACK = 4

# The ZERO WIDTH JOINER codepoint.
ZWJ = 0x200D


# ─────────────────────────────────────────────────────────────────────
# §2 Sub-threat ADT + verdict
# ─────────────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class CombiningStackOverflow:
    """A combining-mark stack of ``stack_len`` marks on the base at
    ``base_pos`` (``stack_len >= MIN_COMBINING_STACK``)."""

    base_pos: int
    stack_len: int


@dataclass(frozen=True, slots=True)
class VariationSelectorVariance:
    """A variation selector at ``first_vs_pos`` (codepoint
    ``first_vs_cp``)."""

    first_vs_pos: int
    first_vs_cp: int


@dataclass(frozen=True, slots=True)
class UnregisteredZwjVariance:
    """A ZWJ-containing input not present in the registered RGI ZWJ set;
    ``first_zwj_pos`` is the position of the first ZWJ."""

    first_zwj_pos: int


@dataclass(frozen=True, slots=True)
class FullwidthVariance:
    """A fullwidth/halfwidth codepoint at ``first_fw_pos`` (codepoint
    ``first_fw_cp``)."""

    first_fw_pos: int
    first_fw_cp: int


@dataclass(frozen=True, slots=True)
class MixedDirectionVariance:
    """Both strong-LTR and strong-RTL codepoints in one input."""

    ltr_count: int
    rtl_count: int


SubThreat = Union[
    CombiningStackOverflow,
    VariationSelectorVariance,
    UnregisteredZwjVariance,
    FullwidthVariance,
    MixedDirectionVariance,
]


def sub_threat_tag(sub: SubThreat) -> str:
    """Fixture-row tag string for a sub-threat (matches ``SubThreat.tag``
    in the Lean/Rust reference)."""
    if isinstance(sub, CombiningStackOverflow):
        return "CombiningStackOverflow"
    if isinstance(sub, VariationSelectorVariance):
        return "VariationSelectorVariance"
    if isinstance(sub, UnregisteredZwjVariance):
        return "UnregisteredZwjVariance"
    if isinstance(sub, FullwidthVariance):
        return "FullwidthVariance"
    if isinstance(sub, MixedDirectionVariance):
        return "MixedDirectionVariance"
    raise TypeError(f"sub_threat_tag: unknown SubThreat variant {sub!r}")


@dataclass(frozen=True, slots=True)
class Classification:
    """Top-level RendererDivergence classification. ``is_clear``
    distinguishes the ``Clear`` variant from the ``Hazard`` variant; a
    hazard carries its sub-threat, the codepoint positions it implicates,
    and the decoded-byte projection (always empty here, kept for shape
    parity with the Lean ``Classification.hazard``)."""

    is_clear: bool
    sub: SubThreat | None = None
    positions: list[int] = field(default_factory=list)
    decoded: list[int] = field(default_factory=list)

    @property
    def tag(self) -> str | None:
        """Human-facing tag for a hazard, or ``None`` when clear."""
        if self.is_clear or self.sub is None:
            return None
        return sub_threat_tag(self.sub)

    @property
    def kind(self) -> ClassificationKind:
        """The classification kind (``CLEAR`` or ``HAZARD``)."""
        return (
            ClassificationKind.CLEAR
            if self.is_clear
            else ClassificationKind.HAZARD
        )


@dataclass(frozen=True, slots=True)
class Verdict:
    """The structured output of :func:`detect` (mirrors the Lean
    ``Verdict``)."""

    input: list[int]
    classify: Classification
    vs_count: int
    combining_count: int
    fullwidth_count: int
    has_zwj: bool
    strong_ltr_count: int
    strong_rtl_count: int


# ─────────────────────────────────────────────────────────────────────
# §3 Core predicates
# ─────────────────────────────────────────────────────────────────────


def is_variation_selector(cp: int) -> bool:
    """True iff ``cp`` is a variation selector (reuses the port's own
    :func:`variation_selector_payload.is_variation_selector`)."""
    return variation_selector_payload.is_variation_selector(cp)


def is_zwj(cp: int) -> bool:
    """True iff ``cp`` is the ZWJ codepoint."""
    return cp == ZWJ


def is_fullwidth_halfwidth(cp: int) -> bool:
    """True iff ``cp`` is in the Halfwidth/Fullwidth Forms block."""
    return 0xFF01 <= cp <= 0xFFEF


def is_grapheme_extend(cp: int) -> bool:
    """True iff ``cp`` has ``Grapheme_Cluster_Break = Extend`` (reuses the
    port's grapheme table via :func:`grapheme.is_grapheme_extend`)."""
    return grapheme.is_grapheme_extend(cp)


# ─────────────────────────────────────────────────────────────────────
# §4 Sub-detectors
# ─────────────────────────────────────────────────────────────────────


def _count_vs(input_cps: list[int]) -> int:
    return sum(1 for cp in input_cps if is_variation_selector(cp))


def _count_combining(input_cps: list[int]) -> int:
    return sum(1 for cp in input_cps if is_grapheme_extend(cp))


def _count_fullwidth(input_cps: list[int]) -> int:
    return sum(1 for cp in input_cps if is_fullwidth_halfwidth(cp))


def _input_has_zwj(input_cps: list[int]) -> bool:
    return any(is_zwj(cp) for cp in input_cps)


def _count_strong_ltr(input_cps: list[int]) -> int:
    return sum(1 for cp in input_cps if ucd.is_strong_ltr(cp))


def _count_strong_rtl(input_cps: list[int]) -> int:
    return sum(1 for cp in input_cps if ucd.is_strong_rtl(cp))


def _first_vs_pos(input_cps: list[int]) -> tuple[int, int] | None:
    """Position and codepoint of the first variation selector."""
    for index, cp in enumerate(input_cps):
        if is_variation_selector(cp):
            return (index, cp)
    return None


def _first_zwj_pos(input_cps: list[int]) -> int | None:
    """Position of the first ZWJ."""
    for index, cp in enumerate(input_cps):
        if is_zwj(cp):
            return index
    return None


def _first_fullwidth_pos(input_cps: list[int]) -> tuple[int, int] | None:
    """Position and codepoint of the first fullwidth/halfwidth codepoint."""
    for index, cp in enumerate(input_cps):
        if is_fullwidth_halfwidth(cp):
            return (index, cp)
    return None


def _first_combining_stack(
    input_cps: list[int], min_stack: int
) -> tuple[int, int] | None:
    """The first base position (a non-Extend codepoint) immediately
    followed by exactly ``min_stack`` consecutive Extend codepoints.
    Returns ``(base_pos, min_stack)`` on hit."""
    for idx, cp in enumerate(input_cps):
        if not is_grapheme_extend(cp):
            following = input_cps[idx + 1 : idx + 1 + min_stack]
            if len(following) == min_stack and all(
                is_grapheme_extend(c) for c in following
            ):
                return (idx, min_stack)
    return None


# ─────────────────────────────────────────────────────────────────────
# §5 Top-level detection
# ─────────────────────────────────────────────────────────────────────


def detect(input_cps: list[int]) -> Verdict:
    """The RendererDivergence detection function."""
    vs_count = _count_vs(input_cps)
    combining_count = _count_combining(input_cps)
    fullwidth_count = _count_fullwidth(input_cps)
    has_zwj = _input_has_zwj(input_cps)
    ltr_count = _count_strong_ltr(input_cps)
    rtl_count = _count_strong_rtl(input_cps)

    # Priority 1: combining-mark stack overflow (Zalgo).
    stack = _first_combining_stack(input_cps, MIN_COMBINING_STACK)
    if stack is not None:
        base_pos, stack_len = stack
        classification = Classification(
            is_clear=False,
            sub=CombiningStackOverflow(base_pos=base_pos, stack_len=stack_len),
            positions=[base_pos],
        )
    else:
        # Priority 2: any variation selector triggers presentation variance.
        vs_hit = _first_vs_pos(input_cps)
        if vs_hit is not None:
            pos, cp = vs_hit
            classification = Classification(
                is_clear=False,
                sub=VariationSelectorVariance(first_vs_pos=pos, first_vs_cp=cp),
                positions=[pos],
            )
        elif has_zwj and not emoji_zwj_integrity.is_registered_zwj_sequence(
            input_cps
        ):
            # Priority 3: ZWJ-containing input not in the registered RGI set.
            zwj_pos = _first_zwj_pos(input_cps)
            if zwj_pos is not None:
                classification = Classification(
                    is_clear=False,
                    sub=UnregisteredZwjVariance(first_zwj_pos=zwj_pos),
                    positions=[zwj_pos],
                )
            else:
                # Unreachable: has_zwj guarantees a ZWJ position exists.
                classification = Classification(is_clear=True)
        else:
            # Priority 4: fullwidth/halfwidth.
            fw_hit = _first_fullwidth_pos(input_cps)
            if fw_hit is not None:
                pos, cp = fw_hit
                classification = Classification(
                    is_clear=False,
                    sub=FullwidthVariance(first_fw_pos=pos, first_fw_cp=cp),
                    positions=[pos],
                )
            elif ltr_count > 0 and rtl_count > 0:
                # Priority 5: mixed direction.
                classification = Classification(
                    is_clear=False,
                    sub=MixedDirectionVariance(
                        ltr_count=ltr_count, rtl_count=rtl_count
                    ),
                    positions=[],
                )
            else:
                classification = Classification(is_clear=True)

    return Verdict(
        input=list(input_cps),
        classify=classification,
        vs_count=vs_count,
        combining_count=combining_count,
        fullwidth_count=fullwidth_count,
        has_zwj=has_zwj,
        strong_ltr_count=ltr_count,
        strong_rtl_count=rtl_count,
    )

File: ports/ruby/lib/unicode_ruby/security/display/renderer_divergence.rb
# frozen_string_literal: true

require_relative "../covert/variation_selector_payload"
require_relative "../identity/emoji_zwj_integrity"
require_relative "../identity/ucd"
require_relative "../../segmentation/grapheme"

module UnicodeRuby
  module Security
    module Display
      # RendererDivergence — detection of codepoint/sequence shapes known to
      # render differently across font + terminal + browser stacks (the
      # display-layer detector).
      #
      # Byte-faithful port of the verified Rust reference
      # (`ports/rust/src/security/display/renderer_divergence.rs`) and of
      # `Unicode/Security/Display/RendererDivergence.lean`.
      #
      # Threat model.  An adversary crafts content that renders one way in the
      # auditor's renderer (a benign glyph or an empty span) and a different way
      # in the consumer's renderer (a misleading glyph, a wider glyph, or a
      # different sequence).  This is the "fingerprint stability" family: clear
      # inputs render the same across the renderer cohort the Standard documents
      # as stable.
      #
      # What the detector draws.  A heuristic three-value split, surfaced through
      # the universal clear/hazard carrier.  An input is clear when none of the
      # documented variance triggers fire, and otherwise is classified by the
      # first trigger in priority order.  It reuses the port's own tables — the
      # variation-selector set (`Covert::VariationSelectorPayload`), the grapheme
      # Extend class (`Segmentation::Grapheme`), the RGI ZWJ registry
      # (`Identity::EmojiZwjIntegrity`), and strong-bidi classes (`Ucd`) — never a
      # host rendering or shaping library.
      #
      # Sub-threats (priority order):
      #   1. CombiningStackOverflow    Zalgo-like combining-mark stack >= 4 on a
      #                                base.
      #   2. VariationSelectorVariance any variation selector present.
      #   3. UnregisteredZwjVariance   ZWJ-containing input not in the RGI ZWJ
      #                                set.
      #   4. FullwidthVariance         a fullwidth/halfwidth codepoint present.
      #   5. MixedDirectionVariance    both strong-LTR and strong-RTL codepoints.
      module RendererDivergence
        # The combining-mark stack depth (on a single base) at or beyond which
        # the input is treated as a Zalgo-style rendering-variance hazard.
        MIN_COMBINING_STACK = 4

        # The ZERO WIDTH JOINER codepoint.
        ZWJ = 0x200D

        # A sub-threat this detector can fire.  `tag` is the fixture-row / wire
        # tag string; `data` carries the variant-specific position payload
        # (`base_pos` + `stack_len`, `first_vs_pos` + `first_vs_cp`,
        # `first_zwj_pos`, `first_fw_pos` + `first_fw_cp`, or `ltr_count` +
        # `rtl_count`).
        SubThreat = Struct.new(:tag, :data)

        # Top-level classification.  `sub` is nil when clear; `positions` is the
        # codepoint indices the sub-threat implicates (empty when clear);
        # `decoded` is the decoded-byte projection (always empty for this
        # detector, kept for shape parity with the Lean `Classification.hazard`).
        Classification = Struct.new(:sub, :positions, :decoded) do
          def clear?
            sub.nil?
          end

          def tag
            sub.nil? ? nil : sub.tag
          end
        end

        # The structured output of `detect` (mirrors the Lean `Verdict`).
        Verdict = Struct.new(
          :input, :classify, :vs_count, :combining_count, :fullwidth_count,
          :has_zwj, :strong_ltr_count, :strong_rtl_count
        )

        module_function

        # ── Sub-threat constructors ────────────────────────────────────────

        # A combining-mark stack of `stack_len` marks on the base at `base_pos`.
        def combining_stack_overflow(base_pos, stack_len)
          SubThreat.new("CombiningStackOverflow", { base_pos: base_pos, stack_len: stack_len })
        end

        # A variation selector at `first_vs_pos` (codepoint `first_vs_cp`).
        def variation_selector_variance(first_vs_pos, first_vs_cp)
          SubThreat.new("VariationSelectorVariance", { first_vs_pos: first_vs_pos, first_vs_cp: first_vs_cp })
        end

        # A ZWJ-containing input not present in the registered RGI ZWJ set.
        def unregistered_zwj_variance(first_zwj_pos)
          SubThreat.new("UnregisteredZwjVariance", { first_zwj_pos: first_zwj_pos })
        end

        # A fullwidth/halfwidth codepoint at `first_fw_pos` (codepoint
        # `first_fw_cp`).
        def fullwidth_variance(first_fw_pos, first_fw_cp)
          SubThreat.new("FullwidthVariance", { first_fw_pos: first_fw_pos, first_fw_cp: first_fw_cp })
        end

        # Both strong-LTR and strong-RTL codepoints in one input.
        def mixed_direction_variance(ltr_count, rtl_count)
          SubThreat.new("MixedDirectionVariance", { ltr_count: ltr_count, rtl_count: rtl_count })
        end

        # ── Classification constructors ────────────────────────────────────

        def clear
          Classification.new(nil, [], [])
        end

        def hazard(sub, positions, decoded)
          Classification.new(sub, positions, decoded)
        end

        # ── Core predicates (reuse the port's own tables) ──────────────────

        # True iff `cp` is a variation selector (reuses the port's predicate).
        def variation_selector?(cp)
          Covert::VariationSelectorPayload.variation_selector?(cp)
        end

        # True iff `cp` is the ZWJ codepoint.
        def zwj?(cp)
          cp == ZWJ
        end

        # True iff `cp` is in the Halfwidth/Fullwidth Forms block.
        def fullwidth_halfwidth?(cp)
          cp >= 0xFF01 && cp <= 0xFFEF
        end

        # True iff `cp` has Grapheme_Cluster_Break = Extend (reuses the port's
        # UAX #29 grapheme table).
        def grapheme_extend?(cp)
          Segmentation::Grapheme.lookup_gcb(cp) == :extend
        end

        # ── Sub-detectors ──────────────────────────────────────────────────

        def count_vs(input)
          input.count { |cp| variation_selector?(cp) }
        end

        def count_combining(input)
          input.count { |cp| grapheme_extend?(cp) }
        end

        def count_fullwidth(input)
          input.count { |cp| fullwidth_halfwidth?(cp) }
        end

        def input_has_zwj(input)
          input.any? { |cp| zwj?(cp) }
        end

        def count_strong_ltr(input)
          input.count { |cp| Ucd.strong_ltr?(cp) }
        end

        def count_strong_rtl(input)
          input.count { |cp| Ucd.strong_rtl?(cp) }
        end

        # Position and codepoint of the first variation selector, or nil.
        def first_vs_pos(input)
          input.each_index do |idx|
            return [idx, input[idx]] if variation_selector?(input[idx])
          end
          nil
        end

        # Position of the first ZWJ, or nil.
        def first_zwj_pos(input)
          input.each_index do |idx|
            return idx if zwj?(input[idx])
          end
          nil
        end

        # Position and codepoint of the first fullwidth/halfwidth codepoint, or
        # nil.
        def first_fullwidth_pos(input)
          input.each_index do |idx|
            return [idx, input[idx]] if fullwidth_halfwidth?(input[idx])
          end
          nil
        end

        # The first base position (a non-Extend codepoint) immediately followed
        # by exactly `min_stack` consecutive Extend codepoints.  Returns
        # `[base_pos, min_stack]` on hit, nil otherwise.
        def first_combining_stack(input, min_stack)
          input.each_index do |idx|
            next if grapheme_extend?(input[idx])

            following = input[(idx + 1), min_stack] || []
            if following.length == min_stack && following.all? { |c| grapheme_extend?(c) }
              return [idx, min_stack]
            end
          end
          nil
        end

        # ── Top-level detection ────────────────────────────────────────────

        # The RendererDivergence detection function.
        def detect(input)
          vs_count = count_vs(input)
          combining_count = count_combining(input)
          fullwidth_count = count_fullwidth(input)
          has_zwj = input_has_zwj(input)
          ltr_count = count_strong_ltr(input)
          rtl_count = count_strong_rtl(input)

          Verdict.new(
            input.dup, classify(input, has_zwj, ltr_count, rtl_count),
            vs_count, combining_count, fullwidth_count, has_zwj,
            ltr_count, rtl_count
          )
        end

        # The priority ladder, factored out of `detect`.
        def classify(input, has_zwj, ltr_count, rtl_count)
          # Priority 1: combining-mark stack overflow (Zalgo).
          stack = first_combining_stack(input, MIN_COMBINING_STACK)
          unless stack.nil?
            base_pos, stack_len = stack
            return hazard(combining_stack_overflow(base_pos, stack_len), [base_pos], [])
          end

          # Priority 2: any variation selector triggers presentation variance.
          vs = first_vs_pos(input)
          unless vs.nil?
            pos, cp = vs
            return hazard(variation_selector_variance(pos, cp), [pos], [])
          end

          # Priority 3: ZWJ-containing input not in the registered RGI set.
          if has_zwj && !Identity::EmojiZwjIntegrity.registered_zwj_sequence?(input)
            zwj_pos = first_zwj_pos(input)
            return hazard(unregistered_zwj_variance(zwj_pos), [zwj_pos], []) unless zwj_pos.nil?

            return clear
          end

          # Priority 4: fullwidth/halfwidth.
          fw = first_fullwidth_pos(input)
          unless fw.nil?
            pos, cp = fw
            return hazard(fullwidth_variance(pos, cp), [pos], [])
          end

          # Priority 5: mixed direction.
          if ltr_count.positive? && rtl_count.positive?
            return hazard(mixed_direction_variance(ltr_count, rtl_count), [], [])
          end

          clear
        end
      end
    end
  end
end
